### DISCUSSION TITLE
IKEv1 IKE_SA rekey destroys CHILD_SAs

### DISCUSSION THREAD

**Discussion by SirHectorTheBootyBarnacle** [strongswan/strongswan#253] [Q&A]
Hi all,

I have a set up that requires IKEv1 and I'm running into a problem when the
IKEv1 Phase 1 (IKE SA) rekeys.  Phase 1 appears to rekey correctly, but
deletes the Phase 2 SAs.
Based on the following website, IPsec SAs are supposed to be adopted by the
new IKE SA and not recreated:
https://wiki.strongswan.org/projects/strongswan/wiki/ExpiryRekey
*"IKEv1* SAs are also rekeyed/reauthenticated using a make-before-break
scheme, however, only the IKE SA is affected. IPsec SAs are adopted by the
new IKE SA and not recreated."

In this setup the Strongswan (192.19.22.10) is configured as the Responder
and a Cisco IOS (192.19.22.1) device as the Initiator.
The initial connection is established, and the traffic is sent ESP
encapsulated.  The initiator attempts to rekey the IKE SA, and appears to
succeed.
Both the Initiator and the Responder are shown with the new IKE SA SPIs,
but during the IKE SA rekey Strongswan deletes the SAD entries for the
IPsec SAs.

Is this a configuration issue on my part, or is there a setting in charon.conf that I can flip?

Cisco appears to be sending a DELETE message as per
https://tools.ietf.org/html/draft-jenkins-ipsec-rekeying-06#section-3.2.

I've included the 'ipsec statusall' outputs, ipsec.conf, and
charon_debug.log
(I've added charon_debug.log as an attachment)

#####
# ipsec.conf - strongSwan IPsec configuration file

# basic configuration
config setup
    # strictcrlpolicy=yes
    # uniqueids = no

# Add connections here.
conn VPNPeer
        leftfirewall=yes
        keyexchange=ikev1

        # Phase 1 settings
        ikelifetime=24h
        margintime=0
        rekeyfuzz=0%
        lifetime=8h
        ike=aes256-sha1-modp2048 !

        # Phase 2
        esp=aes128-sha1-modp2048 !

        left=192.19.22.10
        right=192.19.22.1

        authby=psk

        type=tunnel

        auto=add

        # Rekeying
        #rekey=no

include /var/lib/strongswan/ipsec.conf.inc

#####
Here are the results from 'ipsec statusall':
Initial connection:
#ipsec statusall

Status of IKE charon daemon (weakSwan 5.8.4, Linux 5.5.0-kali2-amd64,
x86_64):
  uptime: 20 seconds, since Mar 09 13:32:08 2021
  malloc: sbrk 1622016, mmap 0, used 610688, free 1011328
  worker threads: 11 of 16 idle, 5/0/0/0 working, job queue: 0/0/0/0,
scheduled: 3
  loaded plugins: charon aesni aes rc2 sha2 sha1 md5 mgf1 random nonce x509
revocation constraints pubkey pkcs1 pkcs7 pkcs8 pkcs12 pgp dnskey sshkey
pem openssl fips-prf gmp agent xcbc hmac gcm drbg attr kernel-netlink
resolve socket-default connmark stroke updown eap-mschapv2 xauth-generic
counters

Listening IP addresses:
  192.19.22.10
  172.19.22.10
  172.17.0.1
Connections:
     VPNPeer:  192.19.22.10...192.19.22.1  IKEv1
     VPNPeer:   local:  [192.19.22.10] uses pre-shared key authentication
     VPNPeer:   remote: [192.19.22.1] uses pre-shared key authentication
     VPNPeer:   child:  dynamic === dynamic TUNNEL

Security Associations (1 up, 0 connecting):
     VPNPeer[1]: ESTABLISHED 11 seconds ago,
192.19.22.10[192.19.22.10]...192.19.22.1[192.19.22.1]
     VPNPeer[1]: IKEv1 SPIs: e53ec81750a41b9c_i 84f72669eb1b150b_r*,
pre-shared key reauthentication in 23 hours
     VPNPeer[1]: IKE proposal:
AES_CBC_256/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_2048
     VPNPeer{1}:  INSTALLED, TUNNEL, reqid 1, ESP SPIs: c8bf9eca_i
2d5a8f29_o
     VPNPeer{1}:  AES_CBC_128/HMAC_SHA1_96/MODP_2048, 400 bytes_i (4 pkts,
9s ago), 400 bytes_o (4 pkts, 9s ago), rekeying in 7 hours
     VPNPeer{1}:   192.19.22.10/32 === 192.19.22.1/32

After IKE Phase 1 rekey:
#ipsec statusall

Status of IKE charon daemon (weakSwan 5.8.4, Linux 5.5.0-kali2-amd64,
x86_64):
  uptime: 6 minutes, since Mar 09 13:32:08 2021
  malloc: sbrk 1622016, mmap 0, used 640656, free 981360
  worker threads: 11 of 16 idle, 5/0/0/0 working, job queue: 0/0/0/0,
scheduled: 5
  loaded plugins: charon aesni aes rc2 sha2 sha1 md5 mgf1 random nonce x509
revocation constraints pubkey pkcs1 pkcs7 pkcs8 pkcs12 pgp dnskey sshkey
pem openssl fips-prf gmp agent xcbc hmac gcm drbg attr kernel-netlink
resolve socket-default connmark stroke updown eap-mschapv2 xauth-generic
counters

Listening IP addresses:
  192.19.22.10
  172.19.22.10
  172.17.0.1
Connections:
     VPNPeer:  192.19.22.10...192.19.22.1  IKEv1
     VPNPeer:   local:  [192.19.22.10] uses pre-shared key authentication
     VPNPeer:   remote: [192.19.22.1] uses pre-shared key authentication
     VPNPeer:   child:  dynamic === dynamic TUNNEL

Security Associations (1 up, 0 connecting):
     VPNPeer[2]: ESTABLISHED 9 seconds ago,
192.19.22.10[192.19.22.10]...192.19.22.1[192.19.22.1]
     VPNPeer[2]: IKEv1 SPIs: e53ec81702d2cc91_i 47da289647a60462_r*,
pre-shared key reauthentication in 23 hours
     VPNPeer[2]: IKE proposal:
AES_CBC_256/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_2048

[charon_debug.log](https://github.com/strongswan/strongswan/files/6124289/charon_debug.log)

**Comment by Thermi** [comment]
> Tue, 2021-03-09, %H:38:37 07[IKE] <VPNPeer|1> received DELETE for IKE_SA VPNPeer[1]


The other peer deletes the IKE_SA and the CHILD_SA without establishing a new IKE_SA first. Check logs of other peer.

**Reply by SirHectorTheBootyBarnacle** [reply]
Hi Thermi,
Yes, I agree that it sends the IKE DELETE message prior to establishing the IKE_SA.

Can Strongswan be configured to handle this?
Would "Break-before-make" be available for IKEv1.  Currently the documentation says it only applies to 


Logs from Cisco:
Mar 11 13:28:19 40.19.22.1 <191>14428938: Mar 11 2021 10:08:52: ISAKMP: (1341):peer does not do paranoid keepalives.

Mar 11 13:28:19 40.19.22.1 <191>14428939: Mar 11 2021 10:08:52: **ISAKMP: (1341):deleting SA reason "IKE SA Lifetime Exceeded" state (I) QM_IDLE       (peer 192.19.22.10)**

Mar 11 13:28:19 40.19.22.1 <191>14428940: Mar 11 2021 10:08:52: IPSEC(key_engine): got a queue event with 1 KMI message(s)

Mar 11 13:28:19 40.19.22.1 <191>14428941: Mar 11 2021 10:08:52: IPSec: Key engine got a KEY_MGR_CHECK_MORE_SAS message

Mar 11 13:28:19 40.19.22.1 <191>14428942: Mar 11 2021 10:08:52: ISAKMP: (1341):set new node 1852108752 to QM_IDLE     

Mar 11 13:28:19 40.19.22.1 <191>14428943: Mar 11 2021 10:08:52: ISAKMP-PAK: (1341):sending packet to 192.19.22.10 my_port 500 peer_port 500 (I) QM_IDLE     

Mar 11 13:28:19 40.19.22.1 <191>14428944: Mar 11 2021 10:08:52: ISAKMP: (1341):Sending an IKE IPv4 Packet.

Mar 11 13:28:19 40.19.22.1 <191>14428945: Mar 11 2021 10:08:52: ISAKMP: (1341):purging node 1852108752

Mar 11 13:28:19 40.19.22.1 <191>14428946: Mar 11 2021 10:08:52: **ISAKMP: (1341):Input = IKE_MESG_INTERNAL, IKE_PHASE1_DEL**

Mar 11 13:28:19 40.19.22.1 <191>14428947: Mar 11 2021 10:08:52: ISAKMP: (1341):Old State = IKE_P1_COMPLETE  New State = IKE_DEST_SA

Mar 11 13:28:19 40.19.22.1 <191>14428948: Mar 11 2021 10:08:52: ISAKMP: (0):SA request profile is (NULL)

Mar 11 13:28:19 40.19.22.1 <191>14428949: Mar 11 2021 10:08:52: ISAKMP: (0):Found a peer struct for 192.19.22.10, peer port 500

Mar 11 13:28:19 40.19.22.1 <191>14428950: Mar 11 2021 10:08:52: ISAKMP: (0):Locking peer struct 0x124211CC, refcount 2 for isakmp_initiator

Mar 11 13:28:19 40.19.22.1 <191>14428951: Mar 11 2021 10:08:52: ISAKMP: (0):local port 500, remote port 500

Mar 11 13:28:19 40.19.22.1 <191>14428952: Mar 11 2021 10:08:52: ISAKMP: (0):Find a dup sa in the avl tree during calling isadb_insert sa = 13BAAC6C

Mar 11 13:28:19 40.19.22.1 <189>14428953: Mar 11 2021 10:08:52: %CRYPTO-5-IKMP_AG_MODE_DISABLED: Unable to initiate or respond to Aggressive Mode while disabled

Mar 11 13:28:19 40.19.22.1 <191>14428954: Mar 11 2021 10:08:52: ISAKMP: (0):Can not start Aggressive mode, trying Main mode.

Mar 11 13:28:19 40.19.22.1 <191>14428955: Mar 11 2021 10:08:52: ISAKMP: (0):found peer pre-shared key matching 192.19.22.10

Mar 11 13:28:19 40.19.22.1 <191>14428956: Mar 11 2021 10:08:52: ISAKMP: (0):constructed NAT-T vendor-rfc3947 ID

Mar 11 13:28:19 40.19.22.1 <191>14428957: Mar 11 2021 10:08:52: ISAKMP: (0):constructed NAT-T vendor-07 ID

Mar 11 13:28:19 40.19.22.1 <191>14428958: Mar 11 2021 10:08:52: ISAKMP: (0):constructed NAT-T vendor-03 ID

Mar 11 13:28:19 40.19.22.1 <191>14428959: Mar 11 2021 10:08:52: ISAKMP: (0):constructed NAT-T vendor-02 ID

Mar 11 13:28:19 40.19.22.1 <191>14428960: Mar 11 2021 10:08:52: ISAKMP: (0):Input = IKE_MESG_FROM_IPSEC, IKE_SA_REQ_MM

Mar 11 13:28:19 40.19.22.1 <191>14428961: Mar 11 2021 10:08:52: ISAKMP: (0):Old State = IKE_READY  New State = IKE_I_MM1

Mar 11 13:28:19 40.19.22.1 <191>14428962: Mar 11 2021 10:08:52: ISAKMP-PAK: (0):sending packet to 192.19.22.10 my_port 500 peer_port 500 (I) MM_NO_STATE

Mar 11 13:28:19 40.19.22.1 <191>14428963: Mar 11 2021 10:08:52: ISAKMP: (0):Sending an IKE IPv4 Packet.

Mar 11 13:28:19 40.19.22.1 <191>14428964: Mar 11 2021 10:08:52: ISAKMP: (1341):deleting SA reason "IKE SA Lifetime Exceeded" state (I) QM_IDLE       (peer 192.19.22.10)

Mar 11 13:28:19 40.19.22.1 <191>14428965: Mar 11 2021 10:08:52: ISAKMP: (0):Unlocking peer struct 0x124211CC for isadb_mark_sa_deleted(), count 1

Mar 11 13:28:19 40.19.22.1 <191>14428966: Mar 11 2021 10:08:52: ISAKMP: (1341):Input = IKE_MESG_FROM_PEER, IKE_MM_EXCH

Mar 11 13:28:19 40.19.22.1 <191>14428967: Mar 11 2021 10:08:52: ISAKMP: (1341):Old State = IKE_DEST_SA  New State = IKE_DEST_SA

Mar 11 13:28:19 40.19.22.1 <191>14428968: Mar 11 2021 10:08:52: ISAKMP-PAK: (0):received packet from 192.19.22.10 dport 500 sport 500 Global (I) MM_NO_STATE

**Reply by Thermi** [reply]
>  reason "IKE SA Lifetime Exceeded"

I'd say you first configure soft limits so the hard limits aren't reached or make the CISCO device rekey the CHILD_SA first.
>     margintime=0
>    rekeyfuzz=0%

That surely doesn't help the situation either.

> Can Strongswan be configured to handle this?
> Would "Break-before-make" be available for IKEv1. Currently the documentation says it only applies to

Only with a monitoring script. And use auto=route/start_action=trap.

**Reply by SirHectorTheBootyBarnacle** [reply]
> > reason "IKE SA Lifetime Exceeded"
> 
> I'd say you first configure soft limits so the hard limits aren't reached or make the CISCO device rekey the CHILD_SA first.

Unfortunately the soft limits are not configurable in the Cisco device.

> margintime=0
> rekeyfuzz=0%
> 
> That surely doesn't help the situation either.

The margintime=0 and rekeyfuzz=0% were set to ensure that the rekeying wasn't done by the Responder (Strongswan) and only by the initiator (Cisco).
Are you seeing something in the logs that is showing the Strongswan closing the IKE_SA prior to the rekey completing?

> Only with a monitoring script. And use auto=route/start_action=trap.

I'll look into this.

Thank you for your help.

**Comment by Thermi** [comment]
There is no indication the CISCO device even attempts a rekeying.

Am March 11, 2021 6:35:51 PM UTC schrieb SirHectorTheBootyBarnacle ***@***.***>:
>> > reason "IKE SA Lifetime Exceeded"
>> 
>> I'd say you first configure soft limits so the hard limits aren't
>reached or make the CISCO device rekey the CHILD_SA first.
>
>Unfortunately the soft limits are not configurable in the Cisco device.
>
>> margintime=0
>> rekeyfuzz=0%
>> 
>> That surely doesn't help the situation either.
>
>The margintime=0 and rekeyfuzz=0% were set to ensure that the rekeying
>wasn't done by the Responder (Strongswan) and only by the initiator
>(Cisco).
>Are you seeing something in the logs that is showing the Strongswan
>closing the IKE_SA prior to the rekey completing?
>
>> Only with a monitoring script. And use auto=route/start_action=trap.
>
>I'll look into this.
>
>Thank you for your help.
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/strongswan/strongswan/discussions/253#discussioncomment-470829

-- 
Sent from mobile

**Reply by SirHectorTheBootyBarnacle** [reply]
> There is no indication the CISCO device even attempts a rekeying.

Could you please explain that?
Are you saying that it appears that of instead of sending an IKE rekey, the Cisco is sending a DELETE then INITIAL_CONTACT messages?
Do you know what messages Strongswan expects to receive to rekey the IKEv1 Phase 1 SA?

I am seeing the following logs:
Tue, 2021-03-09, %H:39:15 05[IKE] <VPNPeer|2> queueing ISAKMP_DELETE task
Tue, 2021-03-09, %H:39:15 05[IKE] <VPNPeer|2> activating new tasks
Tue, 2021-03-09, %H:39:15 05[IKE] <VPNPeer|2>   activating ISAKMP_DELETE task
Tue, 2021-03-09, %H:39:15 05[IKE] <VPNPeer|2> deleting IKE_SA VPNPeer[2] between 192.19.22.10[192.19.22.10]...192.19.22.1[192.19.22.1]
Tue, 2021-03-09, %H:39:15 05[IKE] <VPNPeer|2> sending DELETE for IKE_SA VPNPeer[2]
Tue, 2021-03-09, %H:39:15 05[IKE] <VPNPeer|2> IKE_SA VPNPeer[2] state change: ESTABLISHED => DELETING
Tue, 2021-03-09, %H:39:15 05[ENC] <VPNPeer|2> generating INFORMATIONAL_V1 request 1114736905 [ HASH D ]
Tue, 2021-03-09, %H:39:15 05[NET] <VPNPeer|2> sending packet: from 192.19.22.10[500] to 192.19.22.1[500] (92 bytes)
Tue, 2021-03-09, %H:39:15 05[IKE] <VPNPeer|2> IKE_SA VPNPeer[2] state change: DELETING => DESTROYING
Tue, 2021-03-09, %H:39:15 07[NET] <3> received packet: from 192.19.22.1[500] to 192.19.22.10[500] (244 bytes)
Tue, 2021-03-09, %H:39:15 07[ENC] <3> parsed ID_PROT request 0 [ SA V V V V ]
Tue, 2021-03-09, %H:39:15 07[CFG] <3> looking for an IKEv1 config for 192.19.22.10...192.19.22.1
Tue, 2021-03-09, %H:39:15 07[CFG] <3>   candidate: 192.19.22.10...192.19.22.1, prio 3100
Tue, 2021-03-09, %H:39:15 07[CFG] <3> found matching ike config: 192.19.22.10...192.19.22.1 with prio 3100
Tue, 2021-03-09, %H:39:15 07[IKE] <3> received NAT-T (RFC 3947) vendor ID
Tue, 2021-03-09, %H:39:15 07[IKE] <3> received draft-ietf-ipsec-nat-t-ike-07 vendor ID
Tue, 2021-03-09, %H:39:15 07[IKE] <3> received draft-ietf-ipsec-nat-t-ike-03 vendor ID
Tue, 2021-03-09, %H:39:15 07[IKE] <3> received draft-ietf-ipsec-nat-t-ike-02\n vendor ID
**Tue, 2021-03-09, %H:39:15 07[IKE] <3> 192.19.22.1 is initiating a Main Mode IKE_SA
Tue, 2021-03-09, %H:39:15 07[IKE] <3> IKE_SA (unnamed)[3] state change: CREATED => CONNECTING**

**Comment by Thermi** [comment]
> Do you know what messages Strongswan expects to receive to rekey the IKEv1 Phase 1 SA?

A rekey of a CHILD_SA is basically establishing a second CHILD_SA with the same traffic selector and then deleting the first CHILD_SA. So it's just a CREATE_CHILD_SA message with the normal payloads.

A rekey of an IKE_SA is not possible in IKEv1 because of the standard itself. What instead happens is a new IKE_SA is established without the INITIAL_CONTACT payload, the CHILD_SA is established/migrated, and the old IKE_SA is deleted.

strongSwan is fully standard compliant in, at the very least, this regard. Up to now, there have been no such interoperability issues found with CISCO devices, so there has to be something special in your case. Please review the logs on the CISCO device on why it didn't initiate a rekeying.

**Reply by SirHectorTheBootyBarnacle** [reply]
Hi Thermi,

Thank you for all your help.

So if I'm understanding the charon and Cisco logs, it seems that the Cisco is sending a IKE_SA delete then sending an INITIAL_CONTACT.  Sort of a break-before-make scenario which is not configurable in Strongswan.

**Reply by Thermi** [reply]
No, in fact it just deletes the IKE_SA and then creates a new one. There's no INITIAL_CONTACT notify here (or whatever the IKEv1 analogue might be). Rekeying does not, besides just following the RFCs, require any cooperation or special behaviour by the other peer. There are no rekeying specific things used. As I wrote, the CISCO device needs to attempt to rekey at least once before sending the DELETE. Otherwise you can't even claim it ever tried to rekey.
